static/js/message_events.js

```diff
@@ -512,13 +526,16 @@ export function update_messages(events) {
         // However, we don't need to rerender message_list.narrowed if
         // we just changed the narrow earlier in this function.
         //
-        // TODO: We can potentially optimize this logic to avoid
-        // calling `update_muting_and_rerender` if the muted
+        // We avoid calling `update_muting_and_rerender` if the muted
         // messages would not match the view before or after this
-        // edit.  Doing so could save significant work, since most
+        // edit. Doing so saves significant work, since most
         // topic edits will not match the current topic narrow in
         // large organizations.
```

Comment:
Should I remove this comment altogether? I think the created function name might be self-explanatory.

Reply:
I think this comment can likely be removed, yeah. Or perhaps rewritten to cover the subtle issue I'm about to flag:

We want to rerender the message list if the set of messages in that message list is affected by this topic move. But how do we test that? Looking at just messages that are currently in the message list will be wrong, because it's possible that messages were moved out of the current view.

I think a correct test would be to, at the start of this function, check if any of the moved messages in the event match the the message list (maybe just checking if the IDs are in the `message_list` is correct?); and repeat that here, after we've updated the data structures. If both tests return false, then we know that neither before or after this edit did any of the affected messages appear in our view, and we can skip the rerender here. 

There may also be a quirk that I'm not sure the `changed_narrow` check here is complete; since I think `changed_narrow` is false in the code paths above where we do `message_lists.current.remove_and_rerender`. Hmm.

Reply:
It seems I have misinterpreted how the topic edit works as not all messages of the same topic are moved when you do it, so now I agree with your suggested approach. 

Regarding the `message_lists.current.remove_and_rerender` call, do we really need to rerender at this point? I guess the `update_muting_and_rerender` call is done later on the same conditions anyway.

Reply:
> It seems I have misinterpreted how the topic edit works as not all messages of the same topic are moved when you do it, so now I agree with your suggested approach.

Yeah, that's an important detail about the model! We should perhaps figure out a place to document that -- any ideas where you'd have read it while getting started on this?

> Regarding the message_lists.current.remove_and_rerender call, do we really need to rerender at this point? I guess the update_muting_and_rerender call is done later on the same conditions anyway.

I think it's unlikely we should need to rerender anything twice in a given view when moving messages. I think things are a little confused here. This function is necessary to actually changed the data state of whether messages that no longer match it (say, you moved them out of the stream for the current stream narrow); and then it also does a rerender.

```
remove_and_rerender(message_ids) {
    this.data.remove(message_ids);
    this.rerender();              
}                                 
```

I suspect that only the `this.data.remove(message_ids);` part is necessary here; there's even a comment stating that explicitly:

```
// Ensure messages that are no longer part of this       
// narrow are deleted and messages that are now part     
// of this narrow are added to the message_list.         
//                                                       
// Even if we end up renarrowing, the message_list_data  
// part of this is important for non-rendering message   
// lists, so we do this unconditionally.  Most correctly,
// this should be a loop over all valid message_list_data
// objects, without the rerender (which will naturally   
// happen in the following code).                        
```

This may require a bit more involved of a refactor to do well. But I think probably a good strategy would involve merging these two blocks; ideally with the final version being a loop over `message_lists.all_rendered_message_lists` -- we are working towards having a more generic implementation of supporting multiple simultaneously rendered message lists (not just `narrowed_msg_list` and `home_msg_list`), and this is probably one of the most thorny things to change to achieve that.